src/hotspot/share/gc/g1/g1HeapSizingPolicy.cpp

```diff
@@ -38,162 +38,307 @@ G1HeapSizingPolicy* G1HeapSizingPolicy::create(const G1CollectedHeap* g1h, const
+//                           Steepness at 10 is nearly a step function.
+//
+// In this case, we choose a steepness of 6.0:
+// - For small deviations, the sigmoid output is close to 0, resulting in scale factors near the
+//   lower bound, preventing excessive resizing.
+// - As cpu_usage_delta grows toward 1.0, the steepness value makes the transition sharper, enabling
+//   more aggressive scaling for large deviations.
+//
+// This helps avoid overreacting to small gc_cpu_usage deviations but respond appropriately
+// when necessary.
+double G1HeapSizingPolicy::scale_cpu_usage_delta(double cpu_usage_delta,
+                                                 double min_scale_factor,
+                                                 double max_scale_factor) const {
+  double sigmoid = sigmoid_function(cpu_usage_delta);
+
+  double scale_factor = min_scale_factor + (max_scale_factor - min_scale_factor) * sigmoid;
+  return scale_factor;
+}
 
-  size_t expand_bytes = 0;
+// Calculate the relative difference between a and b.
+static double rel_diff(double a, double b) {
+  return (a - b) / b;
+}
 
-  if (_g1h->capacity() == _g1h->max_capacity()) {
-    log_expansion(short_term_pause_time_ratio, long_term_pause_time_ratio,
-                  threshold, pause_time_threshold, true, 0);
-    clear_ratio_check_data();
-    return expand_bytes;
+size_t G1HeapSizingPolicy::young_collection_expand_amount(double cpu_usage_delta) const {
+  assert(cpu_usage_delta >= 0.0, "must be");
+
+  size_t reserved_bytes = _g1h->max_capacity();
+  size_t committed_bytes = _g1h->capacity();
+  size_t uncommitted_bytes = reserved_bytes - committed_bytes;
+  size_t expand_bytes_via_pct = uncommitted_bytes * G1ExpandByPercentOfAvailable / 100;
+  size_t min_expand_bytes = MIN2(G1HeapRegion::GrainBytes, uncommitted_bytes);
+
+  // Take the current size or G1ExpandByPercentOfAvailable % of
+  // the available expansion space, whichever is smaller, as the base
+  // expansion size. Then possibly scale this size according to how much the
+  // GC CPU usage (on average) has exceeded the target.
+  const double min_scale_factor = 0.2;
+  const double max_scale_factor = 2.0;
+
+  double scale_factor = scale_cpu_usage_delta(cpu_usage_delta,
+                                              min_scale_factor,
+                                              max_scale_factor);
+
+  size_t resize_bytes = MIN2(expand_bytes_via_pct, committed_bytes);
+
+  resize_bytes = static_cast<size_t>(resize_bytes * scale_factor);
+
+  // Ensure the expansion size is at least the minimum growth amount
+  // and at most the remaining uncommitted byte size.
+  return clamp(resize_bytes, min_expand_bytes, uncommitted_bytes);
+}
+
+size_t G1HeapSizingPolicy::young_collection_shrink_amount(double cpu_usage_delta, size_t allocation_word_size) const {
+  assert(cpu_usage_delta >= 0.0, "must be");
+
+  const double min_scale_factor = G1ShrinkByPercentOfAvailable / 1000.0;
```

Comment:
I wonder if this can be written as `min = max / 10`. At first glance, I thought it's a typo, `1000` instead of `100`, but the key msg is that min is 1/10 of max (next line), right?